filtering.py

```diff
@@ -0,0 +1,574 @@
+        STFT of the mixture signal.
+
+    iterations: int [scalar]
+        number of iterations for the EM algorithm
+
+    use_softmask: boolean
+        * if `False`, then the mixture phase will directly be used with the
+          spectrogram as initial estimates.
+
+        * if `True`, a softmasking strategy will be used as described in
+          :func:`softmask`.
+
+    residual: str or None
+        if not None, an additional target is created, which is
+        equal to the mixture minus the other targets, before application of
+        expectation maximization
+
+    eps: {None, float}
+        Epsilon value to use for computing the separations. This is used
+        whenever division with a model energy is performed, i.e. when
+        softmasking and when iterating the EM.
+        It can be understood as the energy of the additional white noise
+        that is taken out when separating.
+        If `None`, the default value is taken as `torch.finfo(x.dtype).eps`.
+
+    Returns
+    -------
+
+    y: torch.Tensor
+            [complex, shape=(nb_frames, nb_bins, nb_channels, 2, nb_sources)]
+        STFT of estimated sources
+
+    Note
+    ----
+
+    * Be careful that you need *magnitude spectrogram estimates* for the
+      case `softmask==False`.
+    * `softmask=False` is recommended
+    * The epsilon value will have a huge impact on performance. If it's large,
+      only the parts of the signal with a significant energy will be kept in
+      the sources. This epsilon then directly controls the energy of the
+      reconstruction error.
+
+    Warning
+    -------
+    As in :func:`expectation_maximization`, we recommend converting the
+    mixture `x` to double precision `torch.float64` *before* calling
+    :func:`wiener`.
+
+    """
+    if use_softmask:
+        y = softmask(targets_spectrograms, mix_stft, eps=eps)
+    else:
+        # first estimates are obtained by just multiplying with mix phase
+        # we tacitly assume that we have magnitude estimates.
+        angle = torch.atan2(mix_stft[..., 1], mix_stft[..., 0])[..., None]
+        nb_sources = targets_spectrograms.shape[-1]
+        y = torch.zeros(mix_stft.shape + (nb_sources,), dtype=mix_stft.dtype,
+                        device=mix_stft.device)
+        y[..., 0, :] = targets_spectrograms * torch.cos(angle)
+        y[..., 1, :] = targets_spectrograms * torch.sin(angle)
+
+    if residual is not None:
+        # if required, adding an additional target as the mix minus
+        # available targets
+        y = torch.cat(
+            [y, mix_stft[..., None] - y.sum(dim=-1, keepdim=True)],
+            dim=-1
+        )
+
+    if not iterations:
+        return y
+
+    # we need to refine the estimates. Scales down the estimates for
+    # numerical stability
+    max_abs = torch.max(
+        torch.as_tensor(1., dtype=mix_stft.dtype, device=mix_stft.device),
+        torch.sqrt(_norm(mix_stft)).max()/scale_factor
```

Comment:
I know that I asked for this before, but can we 

* either remove this, or 
* make `scale_factor` depend on `eps`? ... e.g. 1/eps?